### PR TITLE
commiting DIO Driver includeing video

### PR DESCRIPTION
**Pull request by AhmedEssamMo**

### CODE REVIEW

**Review by MohSamyN — APPROVED** [review]
Excellent Job. Neat and clean.

File: LED_BUTTONS/Button.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include"Macros.h"
+#include"TypeDef.h"
+#include"Button_retval.h"
+#include"DIO_retval.h"
+#include"DIO_REG.h"
+#include"DIO.h"
+#include<util/delay.h>
+
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:8]: No brief comment to define what is the usage of this function.

File: LED_BUTTONS/Button.c [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+Button_ERROR_state_t Button_Init(uint8_t PortID, uint8_t PinNumber) {
+	uint8_t state = 0;
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:10]: You are not following Sprints_CodeGuideLine 


[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/Button.c (hunk 1 above)
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:8]: No brief comment to define what is the usage of this function.

File: LED_BUTTONS/Button.c (hunk 2 above)
Comment (AsaadKam) [reply LED_BUTTONS/Button.c:10]: i meant for uint8_t state = 0;

File: LED_BUTTONS/Button.c [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		return E_Button_SUCCESS ;
+
+	}
+}
+uint8_t Button_getState(uint8_t PortID, uint8_t PinNumber) {
+	uint8_t currentValue = 0;
+	uint8_t state=0;
+	uint8_t oldValue = 0;
+	volatile uint8_t counter = 0;
+	volatile uint32_t Delay = 0;
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:27]: You are not following Sprints_CodeGuideLine 


[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/Button.h [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#ifndef Button_H_
+#define Button_H_
+
+uint8_t Button_getState(uint8_t PortID, uint8_t PinNumber);
+Button_ERROR_state_t Button_Init(uint8_t PortID, uint8_t PinNumber);
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.h:5]: Description for the prototype is needed

File: LED_BUTTONS/Button_retval.h [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#ifndef Button_retval_H_
+#define Button_retval_H_
+
+typedef uint8_t Button_ERROR_state_t;
+
+#define           E_Button_SUCCESS             (Button_ERROR_state_t)( 0 )
+#define           E_Button_INVALID             (Button_ERROR_state_t)( 1 )
+
+
+
```
Comment (AsaadKam) [inline LED_BUTTONS/Button_retval.h:10]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/Button.c [hunk 6]
```diff
@@ -0,0 +1,38 @@
+#include"Macros.h"
+#include"TypeDef.h"
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:2]: -comments doesn't follow Sprints_CodeGuideLine of c files

[Sprints_CodeGuideLine.pdf](https://github.com/KhalidMosad/Sprints-Tasks/files/6901176/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/Button.h [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+#endif  /*Button_H_*/
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.h:6]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO.c [hunk 8]
```diff
@@ -0,0 +1,312 @@
+#include"Macros.h"
+#include"TypeDef.h"
+#include"DIO_retval.h"
+#include"DIO_REG.h"
+#include"DIO.h"
+
+
+
+/* APIs IMPLEMENTATION */
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:10]: -comments doesn't follow Sprints_CodeGuideLine of c files

[Sprints_CodeGuideLine.pdf](https://github.com/KhalidMosad/Sprints-Tasks/files/6901176/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/Button_retval.h [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#ifndef Button_retval_H_
+#define Button_retval_H_
+
+typedef uint8_t Button_ERROR_state_t;
+
+#define           E_Button_SUCCESS             (Button_ERROR_state_t)( 0 )
+#define           E_Button_INVALID             (Button_ERROR_state_t)( 1 )
```
Comment (AsaadKam) [inline LED_BUTTONS/Button_retval.h:7]: Error handling  does not match the formula below




(https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6901387/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO_retval.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ *     This Error Happens if PIN_DIRECTION is not INPUT OR OUTPUT
+ */
+#define           E_DIO_INVALID_PIN_DIRECTION        (DIO_ERROR_state_t)( 2 )
+
+/*
+ *     This Error Happens if setting PUR to non input pin
+ */
+#define           E_DIO_SET_TO_NON_INPUT        (DIO_ERROR_state_t)( 3 )
+
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO_retval.h:25]: Error handling  does not match the formula below




(https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6901387/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO_retval.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
+#ifndef DIO_retval_H_
+#define DIO_retval_H
+
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO_retval.h:4]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO_REG.h [hunk 12]
```diff
@@ -0,0 +1,22 @@
+#ifndef DIO_REG_H_
+#define DIO_REG_H_
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO_REG.h:3]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf]

(https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO.h [hunk 13]
```diff
@@ -0,0 +1,45 @@
+#ifndef _DIO_H_
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.h:1]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/DIO.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ * APIs Declaration
+ */
+DIO_ERROR_state_t DIO_setPortDirection(uint8_t PORTID, uint8_t PortDirection);
+DIO_ERROR_state_t DIO_setPinDirection(uint8_t PortID , uint8_t PinNumber, uint8_t PinDirection);
+DIO_ERROR_state_t DIO_writePortValue(uint8_t PortID, uint8_t PortValue);
+DIO_ERROR_state_t DIO_writePinValue(uint8_t PortID, uint8_t PinNumber,uint8_t PinDirection);
+DIO_ERROR_state_t DIO_readPort(uint8_t PortID, uint8_t* PortState);
+DIO_ERROR_state_t DIO_readPin(uint8_t PortID,uint8_t PinNumber, uint8_t* PinState);
+DIO_ERROR_state_t DIO_TogglePin(uint8_t PortID, uint8_t PinNumber);
+DIO_ERROR_state_t DIO_SetPullupRes(uint8_t PortID, uint8_t PinNumber);
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.h:44]: -Description for the prototype is needed

File: LED_BUTTONS/DIO.c [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+			PORTB_DIR = PortDirection;
+			break;
+		case PORTC :
+			PORTC_DIR = PortDirection;
+			break;
+		case PORTD :
+			PORTD_DIR = PortDirection;
+			break;
+
+		} //END of SWITCH
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:29]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+			case PORTB :
+				Set_Bit(PORTB_DIR, PinNumber);
+				break;
+			case PORTC :
+				Set_Bit(PORTC_DIR, PinNumber);
+				break;
+			case PORTD :
+				Set_Bit(PORTD_DIR, PinNumber);
+				break;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:58]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+			case PORTB :
+				Clr_Bit(PORTB_DIR, PinNumber);
+				break;
+			case PORTC :
+				Clr_Bit(PORTC_DIR, PinNumber);
+				break;
+			case PORTD :
+				Clr_Bit(PORTD_DIR, PinNumber);
+				break;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:76]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/Button.c [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+Button_ERROR_state_t Button_Init(uint8_t PortID, uint8_t PinNumber) {
+	uint8_t state = 0;
+	state = DIO_setPinDirection(PortID, PinNumber, PIN_INPUT);
+	state = DIO_SetPullupRes(PortID, PIN_INPUT);
+	if (state > E_DIO_SUCCESS) {
+		return E_Button_INVALID ;
+	}
+
+	else {
+		return E_Button_SUCCESS ;
```
Comment (AsaadKam) [inline LED_BUTTONS/Button.c:18]: -A function should have a single point of exit at the end is not preferred according to misra rules

https://pvs-studio.com/en/docs/warnings/v2506/

File: LED_BUTTONS/DIO.c [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+		case PORTD :
+			PORTD_DIR = PortDirection;
+			break;
+
+		} //END of SWITCH
+		return E_DIO_SUCCESS ;
+	}	  //END of IF
+	else {
+		return E_DIO_INVALID_PORT_NUMBER ;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:34]: -A function should have a single point of exit at the end is not preferred according to misra rules

https://pvs-studio.com/en/docs/warnings/v2506/

File: LED_BUTTONS/DIO.c [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+			return E_DIO_SUCCESS ;
+		} // END of ELSE IF that check for the direction if it is INPUT
+		else {
+			return E_DIO_INVALID_PIN_DIRECTION ;
+		} //END OF ELSE
+
+	}	  //END of IF That checking the port and pin number
+
+	else {
+		return E_DIO_INVALID_PORT_NUMBER ;
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:87]: -A function should have a single point of exit at the end is not preferred according to misra rules

https://pvs-studio.com/en/docs/warnings/v2506/

File: LED_BUTTONS/DIO.c [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+			break;
+		case PORTD :
+			PORTD_DATA = PortValue;
+			break;
+
+		} //END of SWITCH
+		return E_DIO_SUCCESS ;
+	}	  //END of IF
+	else {
+		return E_DIO_INVALID_PORT_NUMBER ;
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:114]: -A function should have a single point of exit at the end is not preferred according to misra rules

https://pvs-studio.com/en/docs/warnings/v2506/

File: LED_BUTTONS/DIO.c [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+			break;
+		case PORTB :
+			PORTB_DATA = PortValue;
+			break;
+		case PORTC :
+			PORTC_DATA = PortValue;
+			break;
+		case PORTD :
+			PORTD_DATA = PortValue;
+			break;
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:108]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+			case PORTB :
+				Set_Bit(PORTB_DATA, PinNumber);
+				break;
+			case PORTC :
+				Set_Bit(PORTC_DATA, PinNumber);
+				break;
+			case PORTD :
+				Set_Bit(PORTD_DATA, PinNumber);
+				break;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:139]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+				break;
+			case PORTB :
+				Clr_Bit(PORTB_DATA, PinNumber);
+				break;
+			case PORTC :
+				Clr_Bit(PORTC_DATA, PinNumber);
+				break;
+			case PORTD :
+				Clr_Bit(PORTD_DATA, PinNumber);
+				break;
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:156]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 25]
```diff
@@ -0,0 +1,312 @@
...
+			break;
+		case PORTB :
+			*(PortState) = PORTB_STAT;
+			break;
+		case PORTC :
+			*(PortState) = PORTC_STAT;
+			break;
+		case PORTD :
+			*(PortState) = PORTD_STAT;
+			break;
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:189]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 26]
```diff
@@ -0,0 +1,312 @@
...
+		case PORTB :
+			*(PinState) = Get_Bit(PORTB_STAT, PinNumber);
+			break;
+		case PORTC :
+			*(PinState) = Get_Bit(PORTC_STAT, PinNumber);
+			break;
+		case PORTD :
+			*(PinState) = Get_Bit(PORTD_STAT, PinNumber);
+			break;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:220]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 27]
```diff
@@ -0,0 +1,312 @@
...
+			case PORTB :
+				Toggle_Bit(PORTB_DATA, PinNumber);
+				break;
+			case PORTC :
+				Toggle_Bit(PORTC_DATA, PinNumber);
+				break;
+			case PORTD :
+				Toggle_Bit(PORTD_DATA, PinNumber);
+				break;
+
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:252]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/DIO.c [hunk 28]
```diff
@@ -0,0 +1,312 @@
...
+			return E_DIO_INVALID_PORT_NUMBER ;
+
+		}	  //END of ELSE
+
+}
+DIO_ERROR_state_t DIO_SetPullupRes(uint8_t PortID, uint8_t PinNumber){
+	//Check for the ports if it is between 0 to 3
+		if ((PortID <= PORTD ) && (PortID >= PORTA ) && (PinNumber >= PIN0 )
+				&& (PinNumber <= PIN7 )) {
+			switch (PortID) {
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:268]: Here you should define the default case to cover the unexpected condition

File: LED_BUTTONS/app.c [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+			PushedButton=Button_getState(PORTA, PIN4);
+			if (PushedButton == 1) {
+				LED_toggle(PORTA, PIN7);
+			}
+		}
+
+
+	}
+
+	return 0;
```
Comment (AsaadKam) [inline LED_BUTTONS/app.c:30]: Description for each step

File: LED_BUTTONS/app.c [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+	LED_Init(PORTA, PIN7);
+
+	uint8_t state = 0;
+	uint8_t PushedButton = 1;
+	state = Button_Init(PORTA, PIN4);
+
+	while (1) {
+		PushedButton=Button_getState(PORTA, PIN4);
+		_delay_ms(20);
+		if (PushedButton == 0) {
```
Comment (AsaadKam) [inline LED_BUTTONS/app.c:20]: condition should be if(0==PushedButton)

File: LED_BUTTONS/app.c [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+			PushedButton=Button_getState(PORTA, PIN4);
+			if (PushedButton == 1) {
+				LED_toggle(PORTA, PIN7);
+			}
```
Comment (AsaadKam) [inline LED_BUTTONS/app.c:24]: condition should be if(1==PushedButton)

File: LED_BUTTONS/DIO.c [hunk 32]
```diff
@@ -0,0 +1,312 @@
...
+
+			} //END of SWITCH
+			return E_DIO_SUCCESS ;
+
+		}	  //END of IF That checking the port and pin number
+
+		else {
+			return E_DIO_INVALID_PORT_NUMBER ;
+
+		}	  //END of ELSE
```
Comment (AsaadKam) [inline LED_BUTTONS/DIO.c:261]: -A function should have a single point of exit at the end is not preferred according to misra rules

File: LED_BUTTONS/LED.c [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+
+
+void LED_Init(uint8_t PortID, uint8_t PinNumber){
+	uint8_t state=0;
+	state=DIO_setPinDirection(PortID,PinNumber,PIN_OUTPUT);
+}
+void LED_toggle(uint8_t PortID, uint8_t PinNumber){
+	uint8_t state=0;
+	state=DIO_TogglePin(PortID,PinNumber);
+}
```
Comment (AsaadKam) [inline LED_BUTTONS/LED.c:15]: -comments doesn't follow Sprints_CodeGuideLine of c files

[Sprints_CodeGuideLine.pdf](https://github.com/KhalidMosad/Sprints-Tasks/files/6901176/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/LED.h [hunk 34]
```diff
@@ -0,0 +1,12 @@
...
+
+
+
+
+
+
+void LED_Init(uint8_t PortID, uint8_t PinNumber);
+void LED_toggle(uint8_t PortID, uint8_t PinNumber);
+
+#endif /*LED_H_*/
```
Comment (AsaadKam) [inline LED_BUTTONS/LED.h:12]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/LED.h [hunk 35]
```diff
@@ -0,0 +1,12 @@
+#ifndef LED_H_
+#define LED_H_
+
+
+
+
+
+
+void LED_Init(uint8_t PortID, uint8_t PinNumber);
+void LED_toggle(uint8_t PortID, uint8_t PinNumber);
```
Comment (AsaadKam) [inline LED_BUTTONS/LED.h:10]: -Description for the prototype is needed

File: LED_BUTTONS/LED.c [hunk 36]
```diff
@@ -0,0 +1,15 @@
+#include"Macros.h"
+#include"TypeDef.h"
+#include"DIO_REG.h"
+#include"DIO_retval.h"
+#include"DIO.h"
+
+
+void LED_Init(uint8_t PortID, uint8_t PinNumber){
```
Comment (AsaadKam) [inline LED_BUTTONS/LED.c:8]: Description for function is needed

File: LED_BUTTONS/LED.c [hunk 37]
```diff
@@ -0,0 +1,15 @@
...
+	uint8_t state=0;
+	state=DIO_setPinDirection(PortID,PinNumber,PIN_OUTPUT);
+}
+void LED_toggle(uint8_t PortID, uint8_t PinNumber){
```
Comment (AsaadKam) [inline LED_BUTTONS/LED.c:12]: Description for function is needed

File: LED_BUTTONS/Macros.h [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+#define Toggle_High_Nibble(reg)     (reg^=((uint8_t)0xf0))
+#define Toggle_Byte(reg)            (reg^=((uint8_t)0xff))
+
+
+/* Get_Bit --> Gets any bit */
+/* To get a bit, We use ( AND ) as ( 1 --> Makes No change ) and ( 0 --> Sets OFF )  */
+#define Get_Bit(reg,bit)            (((reg>>bit)&(uint8_t)(0x01)))
+#define Get_Low_Nibble(reg)         ((reg&(uint8_t)0x0f))
+#define Get_High_Nibble(reg)        ((reg>>(uint8_t)4))
+
```
Comment (AsaadKam) [inline LED_BUTTONS/Macros.h:33]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/TypeDef.h [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+typedef double     f64_t;
+
+#define NULL_PTR ((void*)0);
+
+
+
+
+
+
+#endif /* TypeDef_H_ */
```
Comment (AsaadKam) [inline LED_BUTTONS/TypeDef.h:27]: -comments doesn't follow Sprints_CodeGuideLine of headers



[Sprints_CodeGuideLine.pdf](https://github.com/AhmedEssamMo/Sprints_Wave_9/files/6900596/Sprints_CodeGuideLine.pdf)

File: LED_BUTTONS/app.c [hunk 40]
```diff
@@ -0,0 +1,31 @@
...
+		}
+
+
+	}
+
```
Comment (AsaadKam) [inline LED_BUTTONS/app.c:29]: -comments doesn't follow Sprints_CodeGuideLine of c files

[Sprints_CodeGuideLine.pdf](https://github.com/KhalidMosad/Sprints-Tasks/files/6901176/Sprints_CodeGuideLine.pdf)

**Pull request merged**